File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.py
class TrieNode:
    def __init__(self):
        self.children = [None] * 26 #Array from a - z
        self.endWord = False

class Trie:

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        curr = self.root

        for char in word:
            index = ord(char) - ord('a')

            if curr.children[index] is None:
                curr.children[index] = TrieNode()
            curr = curr.children[index]
        
        curr.endWord = True

    def search(self, word: str) -> bool:
        curr = self.root

        for char in word:
            index = ord(char) - ord('a')

            if curr.children[index] is None:
                return False
            curr = curr.children[index]
        return curr.endWord

    def startsWith(self, prefix: str) -> bool:
        curr = self.root

        for char in prefix:
            index = ord(char) - ord('a')
            if curr.children[index] is None:
                return False
            curr = curr.children[index]
        
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.cpp
class Trie {


private:

    struct Node 
    {
        bool isWord;
        unordered_map<char, Node*> children;

        //Constructor
        Node() : isWord(false) {}
    };

    Node* root; //Reference to our root.

public:
    
    Trie() {
        root = new Node();
    }
    
    void insert(string word) {
        
        //1. Insert a word. 
        //2. Loop through every char
        Node* curr = root;

        for (char c : word)
        {
            if (curr->children.find(c) == curr->children.end())
            {
                //Not found!
                curr->children[c] = new Node();
            }
            //Move to next node..
            curr = curr->children[c];
        }
        curr->isWord = true;
    }
    
    bool search(string word) {
        Node* curr = root;

        for (char c : word)
        {
            if (curr->children.find(c) == curr->children.end())
            {
                return false;
            }
            curr = curr->children[c];
        }

        return curr->isWord;
    }
    
    bool startsWith(string prefix) {
        
        Node* curr = root;

        for (char c : prefix)
        {
            if (curr->children.find(c) == curr->children.end())
            {
                return false;
            }
            curr = curr->children[c];
        }

        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */